onnxruntime/core/providers/qnn/builder/opbuilder/instance_norm_op_builder.cc

```diff
@@ -0,0 +1,136 @@
+  }
+
+  std::vector<uint32_t> input_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[0].node_arg, input_shape), "Cannot get shape of input 0");
+  const size_t input_rank = input_shape.size();
+
+  if (input_rank <= 2 || input_rank > 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm only supports input ranks of size 3 or 4.");
+  }
+
+  if (node_unit.Domain() != kMSInternalNHWCDomain) {  // Layout has not been transformed to NHWC, so do it here.
+    const uint32_t chans = input_shape[1];
+    for (size_t i = 1; i < input_rank - 1; ++i) {
+      input_shape[i] = input_shape[i + 1];
+    }
+
+    input_shape[input_rank - 1] = chans;
+  }
+
+  const uint32_t num_channels = input_shape.back();
+
+  std::vector<uint32_t> scale_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[1].node_arg, scale_shape), "Cannot get shape of input 1 (scale)");
+  if (scale_shape.size() != 1 || scale_shape[0] != num_channels) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm input 1 (scale) must have 1D shape [channel].");
+  }
+
+  std::vector<uint32_t> bias_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[2].node_arg, bias_shape), "Cannot get shape of input 2 (bias)");
+  if (bias_shape.size() != 1 || bias_shape[0] != num_channels) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm input 2 (bias) must have 1D shape [channel].");
+  }
+
+  NodeAttrHelper node_helper(node_unit);
+  const float epsilon = node_helper.Get("epsilon", 0.0f);
+  if (epsilon <= 0.0f) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm epsilon must be greater than 0.0");
+  }
+
+  return Status::OK();
+}
+
+Status InstanceNormOpBuilder::ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper,
+                                                          const NodeUnit& node_unit,
+                                                          std::vector<std::string>&& input_names,
+                                                          const logging::Logger& logger,
+                                                          bool is_quantized_model,
+                                                          bool do_op_validation) const {
+  NodeAttrHelper node_helper(node_unit);
+  std::vector<std::string> param_tensor_names;
+
+  const float epsilon = node_helper.Get("epsilon", 1e-05f);  // Default is 1e-05 according to ONNX spec.
+  Qnn_Scalar_t epsilon_param = QNN_SCALAR_INIT;
+  epsilon_param.dataType = QNN_DATATYPE_FLOAT_32;
+  epsilon_param.floatValue = epsilon;
+  QnnParamWrapper epsilon_param_wrapper(node_unit.Index(),
+                                        node_unit.Name(),
+                                        qnn_def::epsilon,
+                                        epsilon_param);
+  param_tensor_names.push_back(epsilon_param_wrapper.GetParamTensorName());
+  qnn_model_wrapper.AddParamWrapper(std::move(epsilon_param_wrapper));
+
+  output_count_ = 1;
```

Comment:
Why need to explicit specify only 1 output? We don't need to set it normally except a case that Onnx has multiple outputs and Qnn only use the 1st one.

Reply:
I see. Will change.